### PR TITLE
add new function to write external plugin cue schema into virtual fs

### LINKED ISSUE #33899: add method on plugin interface to return cue schema of plugin panels.
Add new method on plugin interface to get cue schema in order to build cue schema of complete dashboard.

### PR DESCRIPTION
**Pull request by ying-jeanne**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #33899

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/schemaloader/schemaloader.go [hunk 1]
```diff
@@ -147,3 +153,101 @@ func removeNilArray(initialArray []interface{}) []interface{} {
...
+	return nil
+}
+
+type InstanceFS struct {
+	files map[string]*file
+}
+
+func NewInstanceFS() *InstanceFS {
+	return &InstanceFS{
+		files: make(map[string]*file),
```
Comment (ying-jeanne) [inline pkg/services/schemaloader/schemaloader.go:213]: this is probably not enough, regarding if we want to reuse https://github.com/grafana/grafana/blob/main/pkg/schema/load/panel.go#L72, need to have directory instead only plain files

File: pkg/services/schemaloader/InstanceFS.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
+package schemaloader
```
Comment (sdboyer) [inline pkg/services/schemaloader/InstanceFS.go:1]: filename should be `instancefs.go`

File: pkg/services/schemaloader/InstanceFS.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+import (
+	"bytes"
```
Comment (sdboyer) [inline pkg/services/schemaloader/InstanceFS.go:4]: looks like we need gofmt

File: pkg/services/schemaloader/InstanceFS.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+func (d *dir) Name() string {
+	return d.name
+}
+
+func (d *dir) Size() int64 {
+	return 0
+}
+
+func (d *dir) Mode() fs.FileMode {
+	return fs.ModeDir | 0444
```
Comment (sdboyer) [inline pkg/services/schemaloader/InstanceFS.go:78]: Why `0444`?

File: pkg/services/schemaloader/instanceFS_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package schemaloader
```
Comment (sdboyer) [inline pkg/services/schemaloader/instanceFS_test.go:1]: nit: filename should be `instancefs_test.go`

**Review by sdboyer — COMMENTED** [review]
Review is partway done, but i looked around and [`fstest.MapFS`](https://tip.golang.org/pkg/testing/fstest/#MapFS). Is there something special about what we're doing here that requires our own fs implementation, rather than just using that one?

**Review by sdboyer — CHANGES_REQUESTED** [review]
whoops, used wrong review state

File: pkg/services/schemaloader/schemaloader.go [hunk 6]
```diff
@@ -23,33 +26,71 @@ func init() {
...
 type SchemaLoaderService struct {
-	log        log.Logger
-	DashFamily schema.VersionedCueSchema
-	Cfg        *setting.Cfg `inject:""`
+	log          log.Logger
+	DashFamily   schema.VersionedCueSchema
+	Cfg          *setting.Cfg `inject:""`
+	pluginFolder string
+	baseLoadPath load.BaseLoadPaths
+}
```
Comment (marefr) [inline pkg/services/schemaloader/schemaloader.go:40]:
```suggestion
	instanceVFS *ScueVFS
}
```

File: pkg/services/schemaloader/schemaloader.go [hunk 7]
```diff
@@ -23,33 +26,71 @@ func init() {
...
+
+type ScueVFS struct {
+	sync.Mutex
+	fs.FS
+	// fstest.MapFS
```
Comment (marefr) [inline pkg/services/schemaloader/schemaloader.go:45]:
```suggestion
	fstest.MapFS
```

File: pkg/services/schemaloader/schemaloader.go [hunk 8]
```diff
@@ -23,33 +26,71 @@ func init() {
...
+func (r *ScueVFS) RemoveInstanceFile(key string) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.FS.(fstest.MapFS), key)
+}
+
+func (rs *SchemaLoaderService) LoadNewPanelPluginSchema(name, content string) error {
+	cueFile := filepath.Join(rs.pluginFolder, name+".cue")
+	rs.log.Info("Write file into virtual file system", "file", cueFile)
+	rs.baseLoadPath.InstanceCueFS.(*ScueVFS).SetInstanceFile(cueFile, content)
```
Comment (marefr) [inline pkg/services/schemaloader/schemaloader.go:63]:
```suggestion
	rs.instanceVFS.SetInstanceFile(cueFile, content)
```

File: pkg/services/schemaloader/schemaloader.go [hunk 9]
```diff
@@ -23,33 +26,71 @@ func init() {
...
+	return nil
+}
+
+func (rs *SchemaLoaderService) RemovePanelPluginSchema(name string) error {
+	cueFile := filepath.Join(rs.pluginFolder, name+".cue")
+	rs.log.Info("Delete file from virtual file system", "file", cueFile)
+	rs.baseLoadPath.InstanceCueFS.(*ScueVFS).RemoveInstanceFile(cueFile)
```
Comment (marefr) [inline pkg/services/schemaloader/schemaloader.go:70]:
```suggestion
	rs.instanceVFS.RemoveInstanceFile(cueFile)
```

File: pkg/services/schemaloader/schemaloader.go [hunk 10]
```diff
@@ -23,33 +26,71 @@ func init() {
...
 
 func (rs *SchemaLoaderService) Init() error {
+	defaultLoadPaths := load.GetDefaultLoadPaths()
+	rs.pluginFolder = filepath.Join("public", "app", "plugins")
+	rs.baseLoadPath = load.BaseLoadPaths{
+		BaseCueFS:       defaultLoadPaths.BaseCueFS,
+		DistPluginCueFS: defaultLoadPaths.DistPluginCueFS,
+		InstanceCueFS:   new(ScueVFS),
+	}
+	rs.baseLoadPath.InstanceCueFS.(*ScueVFS).FS.(fstest.MapFS)[rs.pluginFolder] = &fstest.MapFile{Mode: fs.ModeDir}
```
Comment (marefr) [inline pkg/services/schemaloader/schemaloader.go:82]:
```suggestion
	rs.instanceVFS = &ScueVFS{
		FS: fstest.MapFS{
			rs.pluginFolder: &fstest.MapFile{Mode: fs.ModeDir},
		},
	}
	rs.baseLoadPath = load.BaseLoadPaths{
		BaseCueFS:       defaultLoadPaths.BaseCueFS,
		DistPluginCueFS: defaultLoadPaths.DistPluginCueFS,
		InstanceCueFS:   rs.instanceVFS,
	}
```

**Review by marefr — COMMENTED** [review]
Some suggestions

**Pull request closed without merging**